### PR TITLE
Native assets support for Windows

### PR DESCRIPTION
**Pull request by dcharkes**
Support for FFI calls with `@Native external` functions through Native assets on Windows. This enables bundling native code without any build-system boilerplate code.

For more info see:

* https://github.com/flutter/flutter/issues/129757

### Implementation details for Windows.

Mainly follows the design of https://github.com/flutter/flutter/pull/134031.

Specifically for Windows in this PR is the logic for finding the compiler `cl.exe` and environment variables that contain the paths to the Windows headers `vcvars.bat` based on `vswhere.exe`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dcharkes** [comment]
@stuartmorgan Windows support is also green on CI. Feel free to postpone reviewing until after Linux support lands to not look at the Linux code twice. (This PR is based on top of the Linux PR, and I don't believe GitHub has support for showing the diff between two PRs.)

@stuartmorgan Who should be looped in for the review on Windows?

**Comment by stuartmorgan** [comment]
> @stuartmorgan Who should be looped in for the review on Windows?

I can do the primary review; I'll defer to @cbracken on who on the desktop team is most plugged into the build system.

**Comment by dcharkes** [comment]
This PR is ready for review (CI is green, and parent PR has landed).

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:446]: The "LinuxWindows" naming seems pretty confusing. What is the conceptual thing that makes these the same as each other, but different from iOS/macOS (is it single-arch output files)? That would be a better thing to name it for.

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:478]: This is incorrect, at least for the logging (and if for the code, very confusing, so needs a comment).

Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:478]: `targetOS`

https://dart.dev/effective-dart/style#do-capitalize-acronyms-and-abbreviations-longer-than-two-letters-like-words

File: packages/flutter_tools/lib/src/build_system/targets/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/build_system/targets/native_assets.dart:123]: Why only x64?

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:425]: Why wouldn't we use this for a Windows test host?

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:595]: Why is this Linux/Windows-specific?

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/native_assets.dart:478]: This also suggests there isn't testing of the logging happening for each platform.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:24]: ?

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: Are these paths correct on an arm64 machine, or are there native versions?

File: packages/flutter_tools/templates/app_shared/windows.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/CMakeLists.txt.tmpl:94]: As with Linux, we need a comment on this.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:382]: Is this information not available from `vswhere` output?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/native_assets.dart:446]: I would probably make `*Windows` and `*Linux` versions that are public, and which wrap a private helper named for the conceptual case it handles.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: Do we support arm64?

It doesn't look like it: https://github.com/flutter/flutter/issues/53120

File: packages/flutter_tools/lib/src/build_system/targets/native_assets.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/build_system/targets/native_assets.dart:123]: Sorry, I thought target support (https://github.com/flutter/flutter/issues/129807) had landed recently with the build output structure change, but apparently not yet.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: I believe the host support has landed. Presumably at least the `Hostx64` part of this is the arch of the binary itself, rather than what the target arch is?

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:382]: No, unfortunately not. 😢

https://gist.github.com/dcharkes/7e5d9cc41089e5c8516b27bd2d4baff6

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: @loic-sharma Would you be able to help here?

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/native_assets.dart:441]: Optional: It's weird to have linux/native_assets.dart but for this not to be there; consider making the `*SingleArchitecture` helpers public after all, and then moving the wrappers to the platform-specific native_assets.dart files for consistency.

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/native_assets.dart:626]: What was the answer to this? I'm not sure why a `TargetPlatform`->`Target` mapping function wouldn't cover every supported platform.

File: packages/flutter_tools/lib/src/test/test_compiler.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/test/test_compiler.dart:186]: Can we split this into Linux and Windows cases that call the platform-specific wrappers since the fact that they are currently the same is an implementation detail?

File: packages/flutter_tools/lib/src/windows/native_assets.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/native_assets.dart:27]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-extension

File: packages/flutter_tools/test/general.shard/linux/native_assets_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/linux/native_assets_test.dart:74]: Can these changes be reverted, using the Linux-specific wrapper?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Mostly looks good, but Loic should definitely review as well.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: My Arm64 machine has the following toolchains:

```
VC\Tools\MSVC\<msvc version>\bin\Hostarm64\arm64
VC\Tools\MSVC\<msvc version>\bin\Hostarm64\x64
VC\Tools\MSVC\<msvc version>\bin\Hostx64\x64
VC\Tools\MSVC\<msvc version>\bin\Hostx64\arm64
```

Currently we build x64 executables on Windows Arm64 hosts. These executables run using x64 emulation. We don't support creating Arm64 executables yet, nor do we support cross-compilation. For now, using the x64 toolchain on Arm64 hosts seems reasonable. /cc @pbo-linaro who is working on Windows Arm64 support.

Comment (loic-sharma) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: I ran `test/general.shard/windows/native_assets_test.dart` on my Windows Arm64 machine, all 11 tests passed.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:237]: This exists on my Windows Arm64 machine and targets x64 using the x64 host toolchain. @pbo-linaro, we will need to update this to support targeting Arm64.

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/native_assets.dart:626]: The target locations in MacOS/iOS are overlapping for multiple architectures, because they are fat binaries.
The target locations in Android are also overlapping, but that is because Android automatically resolves within the `jniLibs/lib/$jniArchDir/`.

Maybe I misunderstood the question. Is your question about something else than the paths we embed in the dill file to do lookup at runtime?

Comment (dcharkes) [reply packages/flutter_tools/lib/src/native_assets.dart:626]: Oh, great, yeah GitHub UI is showing your comment on other code, this is probably about the function below.

About the function below. For darwin, ios, and android, we need the DarwinArchs,AndroidArchs information. So for those platforms those environment vars are passed in to the similar function.
For Windows/Linux the architecture is _not_ separate from the target platforms (likely because there is no way to build an app that targets more than one platform). So the information is embedded in the TargetPlatform enum instead of a separate WindowsArchs or LinuxArchs environment variable.

File: packages/flutter_tools/lib/src/windows/native_assets.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/windows/native_assets.dart:27]: Replaced it with an anonymous top level function. Honestly, this is less readable IMO. But happy to stick with with Flutter guidelines if you prefer. (Isn't this a general issue with trying to chain non-methods on nullable things?)

File: packages/flutter_tools/lib/src/linux/build_linux.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/linux/build_linux.dart:49]: The `'linux'` string gets pasted into the Cmake file. The implications of this aren't evident from the call site.

Would it make sense to pass an enum here and have `CmakeNativeAssetsMigration` decide what OS string value should be used when generating the cmake file?

File: packages/flutter_tools/lib/src/macos/native_assets.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/macos/native_assets.dart:43]: Just curious, why the `_` suffix here?

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:382]: Should we return null instead of throwing if there's no msvc version subdirectory?

Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart:382]: Nit: this line is a bit long, consider breaking it up into multiple lines. I don't feel strongly about this.

**Review by loic-sharma — APPROVED** [review]
Windows logic LGTM. The build system changes look reasonable to me as well, but I'm not an expert in that area - I'd consider getting a review from the Flutter tool team.

File: packages/flutter_tools/lib/src/linux/build_linux.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/linux/build_linux.dart:49]: I didn't see anything in `project` or `project.parent` to give me access to the OS.

File: packages/flutter_tools/lib/src/macos/native_assets.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/macos/native_assets.dart:43]: `nativeAssetsBuildUri` used to be called `buildUri` in a previous version. Cleaned up.

**Comment by dcharkes** [comment]
Apart from waiting for FRoB to roll the reland of Linux, this is ready to land.

@stuartmorgan did you want to take another look at it before that?

File: packages/flutter_tools/lib/src/windows/visual_studio.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: I think it's fine to use x64 (hostx64) cl.exe running under emulation, it's always avaiable, and avoid special case regarding host machine.

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: Won't that unnecessarily slow the build process?

Comment (dcharkes) [reply packages/flutter_tools/lib/src/windows/visual_studio.dart:219]: On the M1 Macs the x64 emulation is pretty fast. I have no idea about the quality of x64 emulation on Windows machines.

It's probably cleaner to use the `Hostarm64` if it is available. Though, maybe it's better to do that in a follow up PR by someone who can actually test it, as I can't.

**Comment by dcharkes** [comment]
It's black magic to me when the Google testing decides to work or not.

Anyway, this PR is ready to land if you're happy with it @stuartmorgan.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM

**Pull request merged**
